benchmarks/imagenet/resnet50/dino.py

```diff
@@ -0,0 +1,166 @@
+        teacher_projections = self.projection_head(teacher_features)
+        student_projections = torch.cat(
+            [self.forward_student(global_views), self.forward_student(local_views)]
+        )
+
+        loss = self.criterion(
+            teacher_projections.chunk(2),
+            student_projections.chunk(len(views)),
+            epoch=self.current_epoch,
+        )
+        self.log_dict(
+            {"train_loss": loss, "ema_momentum": momentum},
+            prog_bar=True,
+            sync_dist=True,
+            batch_size=len(targets),
+        )
+
+        # Online classification.
+        cls_loss, cls_log = self.online_classifier.training_step(
+            (teacher_features.chunk(2)[0].detach(), targets), batch_idx
+        )
+        self.log_dict(cls_log, sync_dist=True, batch_size=len(targets))
+        return loss + cls_loss
+
+    def validation_step(
+        self, batch: Tuple[Tensor, Tensor, List[str]], batch_idx: int
+    ) -> Tensor:
+        images, targets = batch[0], batch[1]
+        features = self.forward(images).flatten(start_dim=1)
+        cls_loss, cls_log = self.online_classifier.validation_step(
+            (features.detach(), targets), batch_idx
+        )
+        self.log_dict(cls_log, prog_bar=True, sync_dist=True, batch_size=len(targets))
+        return cls_loss
+
+    def configure_optimizers(self):
+        # Don't use weight decay for batch norm, bias parameters, and classification
+        # head to improve performance.
+        params, params_no_weight_decay = get_weight_decay_parameters(
+            [self.student_backbone, self.student_projection_head]
+        )
+        # For ResNet50 we use SGD instead of AdamW/LARS as recommended by the authors:
+        # https://github.com/facebookresearch/dino#resnet-50-and-other-convnets-trainings
+        optimizer = SGD(
+            [
+                {"name": "dino", "params": params},
+                {
+                    "name": "dino_no_weight_decay",
+                    "params": params_no_weight_decay,
+                    "weight_decay": 0.0,
+                },
+                {
+                    "name": "online_classifier",
+                    "params": self.online_classifier.parameters(),
+                    "weight_decay": 0.0,
+                },
+            ],
+            lr=0.03 * self.batch_size_per_device * self.trainer.world_size / 256,
+            momentum=0.9,
+            weight_decay=1e-4,
+        )
+        scheduler = {
+            "scheduler": CosineWarmupScheduler(
+                optimizer=optimizer,
+                warmup_epochs=(
+                    self.trainer.estimated_stepping_batches
+                    / self.trainer.max_epochs
+                    * 10
+                ),
+                max_epochs=self.trainer.estimated_stepping_batches,
+            ),
+            "interval": "step",
+        }
+        return [optimizer], [scheduler]
+
+    def configure_gradient_clipping(
```

Comment:
where is this coming from? I didn't see anything in the paper

Reply:
Default is 3: https://github.com/facebookresearch/dino/blob/7c446df5b9f45747937fb0d72314eb9f7b66930a/main_dino.py#L87